device.c

```diff
@@ -428,25 +453,52 @@ int adev_open_input_stream(struct audio_hw_device *dev,
         return -EINVAL;
     }
     /* check requested device */
-    /* TODO we should handle only devices described in XML configuration and "default" one */
     if ((devices & AUDIO_DEVICE_BIT_IN) == 0) {
         ALOGE("Failed. Incorrect device type. -EINVAL");
         *stream_in = NULL;
         pthread_mutex_unlock(&xdev->lock);
         return -EINVAL;
     }
 
-    /* TODO define card_id and dev_id for stream, using provided parameters:
-       devices, flags, address, source */
+    unsigned int slot = 0;
+    /* For now we have one card, so card id is always constant */
+    /* Get device id using device type */
+    for (slot = 0; slot < NUMBER_OF_DEVICES_IN; slot++) {
+        if ((devices & xa_input_map[slot].device_type) != 0) {
+            if (xa_input_map[slot].device_type == AUDIO_DEVICE_IN_BUS) {
+                if (sscanf(address, "bus%u", &dev_id) == 1) {
+                    /* dev_id is identified */
+                    dev_id = xa_input_map[slot].pcm_dev_id;
+                } else {
+                    ALOGE("Address has not supported format.");
+                    dev_id = NUMBER_OF_DEVICES_IN;
+                }
+            } else {
+                /* dev_id is identified */
+                dev_id = xa_input_map[slot].pcm_dev_id;
```

Comment:
you use slot to check if dev_id was found, so it is not needed to set dev_id to wrong value here

Reply:
Fixed.